modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/planner/HashAggregatePlannerTest.java

```diff
@@ -42,6 +51,92 @@
+            IgniteDistributions.single(),
+            IgniteDistributions.random(),
+            IgniteDistributions.broadcast(),
+            IgniteDistributions.hash(ImmutableList.of(0, 1, 2, 3)));
+
+        for (IgniteDistribution distr : lst) {
+            TestTable tbl = createTable(distr);
+            publicSchema.addTable("TEST", tbl);
+
+            assertNoIndexCount("SELECT COUNT(*) FROM TEST", publicSchema);
+
+            tbl.addIndex(QueryUtils.PRIMARY_KEY_INDEX, 0);
+
+            assertIndexCount("SELECT COUNT(*) FROM TEST", publicSchema);
+            assertIndexCount("SELECT COUNT(1) FROM TEST", publicSchema);
+            assertIndexCount("SELECT COUNT(*) FROM (SELECT * FROM TEST)", publicSchema);
+            assertIndexCount("SELECT COUNT(1) FROM (SELECT * FROM TEST)", publicSchema);
+
+            assertIndexCount("SELECT COUNT(*), COUNT(*), COUNT(1) FROM TEST", publicSchema);
+
+            // Count on certain fields can't be optimized. Nulls are count included.
+            assertNoIndexCount("SELECT COUNT(VAL0) FROM TEST", publicSchema);
+            assertNoIndexCount("SELECT COUNT(DISTINCT VAL0) FROM TEST", publicSchema);
+
+            assertNoIndexCount("SELECT COUNT(*), COUNT(VAL0) FROM TEST", publicSchema);
+
+            assertNoIndexCount("SELECT COUNT(1), COUNT(VAL0) FROM TEST", publicSchema);
+            assertNoIndexCount("SELECT COUNT(DISTINCT 1), COUNT(VAL0) FROM TEST", publicSchema);
+
+            assertNoIndexCount("SELECT COUNT(*) FILTER (WHERE VAL0>1) FROM TEST", publicSchema);
+
+            // IndexCount can't be used with a condition, groups, other aggregates or distincts.
+            assertNoIndexCount("SELECT COUNT(*) FROM TEST WHERE VAL0>1", publicSchema);
+            assertNoIndexCount("SELECT COUNT(1) FROM TEST WHERE VAL0>1", publicSchema);
+
+            assertNoIndexCount("SELECT COUNT(*), SUM(VAL0) FROM TEST", publicSchema);
+
+            assertNoIndexCount("SELECT VAL0, COUNT(*) FROM TEST GROUP BY VAL0", publicSchema);
+
+            assertNoIndexCount("SELECT COUNT(*) FROM TEST GROUP BY VAL0", publicSchema);
+
+            assertNoIndexCount("SELECT COUNT(*) FILTER (WHERE VAL0>1) FROM TEST", publicSchema);
+
+            publicSchema.addTable("TEST2", createBroadcastTable());
+
+            assertNoIndexCount("SELECT COUNT(*) FROM (SELECT T1.VAL0, T2.VAL1 FROM TEST T1, " +
+                "TEST2 T2 WHERE T1.GRP0 = T2.GRP0)", publicSchema);
+
+            publicSchema.removeTable("TEST");
+        }
+    }
+
+    /** */
+    private void assertIndexCount(String sql, IgniteSchema schema) throws Exception {
+        DistributionFunction distr = ((IgniteTable)schema.getTable("TEST")).distribution().function();
+
+        if (distr.equals(DistributionFunction.hash()) || distr.equals(DistributionFunction.random())) {
+            assertPlan(
+                sql,
+                schema,
+                nodeOrAnyChild(isInstanceOf(IgniteColocatedHashAggregate.class)
```

Comment:
In some cases (depending on cost) plan can be rewriten to `reduce aggregate / exchange / map aggregate / index count` and test can fail in the future. Let's test here just something like for any distribution:
```
nodeOrAnyChild(isInstanceOf(IgniteAggregate.class) 
    .and(nodeOrAnyChild(isInstanceOf(IgniteIndexCount.class))))
```

Reply:
Fixed